src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1913,9 +1939,22 @@ class Simplex_tree {
     return nullptr;
   }
 
-  /** \brief Helper method that returns the corresponding Simplex_handle from a member element defined by a node. */
+  /** \brief Helper method that returns the corresponding Simplex_handle from a member element defined by a node.
+   * Not optimal, if Options::simplex_handle_strong_validity == true. Requires that the node is not a copy.
+   */
   static Simplex_handle simplex_handle_from_node(Node& node) {
-    return (Simplex_handle)(boost::intrusive::get_parent_from_member<Dit_value_t>(&node, &Dit_value_t::second));
+    if constexpr (Options::simplex_handle_strong_validity) {
+      Siblings* children = node.children();
+      //verifies if node is a leaf
+      for (auto it = children->members().begin(); it != children->members().end(); ++it){
+        // WARNING: does not work if nodes were copied, but I don't see another way to verify if two leaf nodes are
+        // equal. And I guess that `get_parent_from_member` needs a real pointer to the flat_map element too anyway?
```

Comment:
I am trying to understand what the problem is here. We have a container (tree or vector) of pairs. From a reference to the second element of the pair, we want an iterator on this pair. get_parent_from_member gets us a pointer to the pair. For the vector iterator, boost kindly provides an explicit constructor of the iterator from the pointer. For a tree, this constructor is missing? I would need to dig through the boost sources to see if there is a workaround. Ideally (?) we would change the coface intrusive list so it contains Simplex_handle instead of `Node*`, but I don't know how hard that is.
Even if we end up using slow code with find, having a linear search before find seems strange.

Reply:
For the map, I was not able to make `get_parent_from_member` work, but I missed perhaps something. The boost documentation is not really helpful... Digging through the sources works perhaps better.
I also don't like having a linear search, but I didn't found another workaround... And replacing Node* with Simplex handles is not really a solution, specifically because they are not always stable. Except if we choose to have two different types of lists, depending on the value of `simplex_handle_strong_validity`, what is clearly possible, just uglier. So, I will look into it.

Reply:
So, the only difficulty there is to replace the Node* by simplex handles is that we either need a wrapper around the handles to inherit the hook or we have to use something else than a intrusive list. The rest is quit straightforward.
If we replace the intrusive list, we need something easy to iterate over and to remove from.

Reply:
> I also don't like having a linear search, but I didn't found another workaround...

Everywhere we call this function, we also know the Vertex_handle that goes with the Node.

> And replacing Node* with Simplex handles is not really a solution, specifically because they are not always stable.

Simplex_handle and Node* are exactly as (un)stable as each other, since they only differ by a constant offset.

Reply:
Boost's map::iterator is roughly of type `iterator_from_iiterator<tree_iterator>`. I can go back and forth between that and a node_ptr (which is a true pointer `compact_rbtree_node*`):
```c++
  auto it = m.begin();
  auto iit = it.get();
  auto pn = iit.pointed_node();
  decltype(iit) iit2; iit2 = pn;
  decltype(it) it2(iit2);
  assert(it==it2);
```
Now going between that node_ptr and our `Node*` (different kind of node), since both are pointers, we can cast to `const char*` and compute the difference to find the offset, and then we can add/remove the offset (and cast twice) to go from one pointer to the other.
Yeah that's pretty ugly, and a bit fragile using public but undocumented stuff, but Boost.Container isn't a very active library.

Reply:
> Everywhere we call this function, we also know the Vertex_handle that goes with the Node.
Ok, if we add the vertex handle in the parameters, we could do the verification in log instead of linear time. If that is fine, I would do that until we have something better.

>  Simplex_handle and Node* are exactly as (un)stable as each other, since they only differ by a constant offset.
Sorry, I was thinking about `Node`, not `Node*`. But isn't that still a problem, by the way? If I remember well, boost intrusive lists do not copy their members, which means that the `link_nodes_by_label` optimization does not work if we add or remove too many things in the tree. Or do references somehow not get invalidated?

Reply:
Look at Hooks_simplex_base_link_nodes: it has a move constructor/assignment to adapt the links any time a node is moved.

Reply:
So, that mean that the list only keeps track of the position of the member `list_max_vertex_hook_` and not of the class `Hooks_simplex_base_link_nodes`? I agree that the hooks are the right ones, but the node it-self and therefore the hook wrapper changed address, no? Or perhaps there is something I am missing about inheritance?

Reply:
I added the vertex handle in the new version. Is it okay this way?
In the meantime, I will be looking at the offset to see if we can switch from Node to iterator as you suggested.

Reply:
So, this works on my side for boost::map :
```C++
static Simplex_handle simplex_handle_from_node(Node& node) {
    const char* shift = (const char*)(&node) - 32;
    boost::intrusive::compact_rbtree_node<void*>* it_ptr = (boost::intrusive::compact_rbtree_node<void*>*)shift;
    boost::intrusive::tree_iterator<
        boost::intrusive::bhtraits<
            boost::container::base_node<
                std::pair<const int, Simplex_tree_node_explicit_storage<Simplex_tree>>,
                boost::container::dtl::intrusive_tree_hook<void*, boost::container::red_black_tree, true>, true>,
            boost::intrusive::rbtree_node_traits<void*, true>, boost::intrusive::normal_link, boost::intrusive::dft_tag,
            3>,
        false>
        it_ii;
    it_ii = it_ptr;
    Dictionary_it sh(it_ii);

    return sh;
  }
```
A more "automatic" version:
```C++
static Simplex_handle simplex_handle_from_node(Node& node) {
    Dictionary_it testIt = node.children()->members().begin();
    Node* testNode = &testIt->second;
    auto testIIt = testIt.get();
    auto testPtr = testIIt.pointed_node();
    auto shift = (const char*)(testNode) - (const char*)(testPtr);
    decltype(testPtr) sh_ptr = decltype(testPtr)((const char*)(&node) - shift);
    decltype(testIIt) sh_ii;
    sh_ii = sh_ptr;
    Dictionary_it sh(sh_ii);

    return sh;
  }
```

Reply:
I know it feels ugly, but it is contained inside this one helper function, so I would be in favor of doing this (say the second "automatic" version). You can add the explicit types as comments next to it if you want.

Reply:
It is also much faster and avoids the label parameter. The new version is committed. I kept the old version as a comment in case some problem arises in the future.